### PR TITLE
Add KVS WebRTC DQP test and metrics

### PR DESCRIPTION
**Pull request by bentcooke**
*Issue #, if available:*

*Description of changes:*
Add option to enable viewer-side test and metrics to support the Amazon Kinesis Video Streams with WebRTC Device Qualification Program.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/index.html [hunk 1]
```diff
@@ -125,6 +126,18 @@ <h4>NAT Traversal</h4>
                 "><sup>&#9432;</sup></span>
                 </div>
             </div>
+            <h4>Amazon KVS WebRTC DQP</h4>
+            <div class="form-group">
+                <div class="form-check form-check-inline">
+                    <input class="form-check-input" type="checkbox" id="enableDQPmetrics" value="enableDQPmetrics">
+                    <label for="enableDQPmetrics" class="form-check-label">Enable KVS WebRTC DQP Test and Metrics (Viewer only)</label>
```
Comment (sirknightj) [inline examples/index.html:133]: Could we make **(viewer only)** in `<small>`

File: examples/index.html [hunk 2]
```diff
@@ -195,6 +208,26 @@ <h5>From Master</h5>
...
+                        <div id="dqp-test"></div>
+                    </div>
+                </div>
+                <div class="col">
+                    <div class="card bg-light mb-3">
+                        <canvas id="metricsChart" style="width:100%"; height="400px"></canvas>
+                    </div>
+                </div>
+            </div>
+            <h3 id="logs-header">Live Stats (from Master)</h3>
```
[the remark this answers is not in the document]
Comment (bentcooke) [reply examples/index.html]: Good catch... fixed!

File: examples/viewer.js [hunk 3]
```diff
@@ -264,3 +365,252 @@ function sendViewerMessage(message) {
...
+                    audioBitrate = 8 * (bytes - aBytesPrev) / (report.timestamp - aTimeStampPrev);
+                    audioBitrate = Math.floor(audioBitrate);
+                }
+                aBytesPrev = bytes;
+                aTimeStampPrev = report.timestamp;
+
+            }
+        });
+
+        // console.log('[DQP TEST] Connected to: ' + connectionString);
```
[the remark this answers is not in the document]
Comment (bentcooke) [reply examples/viewer.js]: removed!

File: examples/viewer.js [hunk 4]
```diff
@@ -264,3 +365,252 @@ function sendViewerMessage(message) {
...
+    let activeCandidatePair = null;
+    let remoteCandidate = null;
+    let connectionString = '';
+    let html_str = '';
+
+    //Loop through each report and find the active pair.
+    stats.forEach(report => {
+        if (report.type == 'transport') {
+            activeCandidatePair = stats.get(report.selectedCandidatePairId);
+            //console.log('[DQP TEST] transport report:',report);
```
[the remark this answers is not in the document]
Comment (bentcooke) [reply examples/viewer.js]: removed!

File: examples/viewer.js [hunk 5]
```diff
@@ -264,3 +365,252 @@ function sendViewerMessage(message) {
         return false;
     }
 }
+
+// Functions to support DQP test and metrics
+function calcDiffTimestamp2Sec(large, small) {
+    var diffMs = (large - small);
+    var num = Number.parseFloat(diffMs).toFixed(2);
+    var diffSec = Number.parseFloat(num / 1000).toFixed(2);
+    return diffSec;
```
[the remark this answers is not in the document]
Comment (bentcooke) [reply examples/viewer.js]: that is a better solution and has been implemented with the last push.

File: examples/viewer.js [hunk 6]
```diff
@@ -264,3 +365,252 @@ function sendViewerMessage(message) {
...
+
+    let rttSum = 0;
+    let vjitterSum = 0;
+    let ajitterSum = 0;
+    let framerateSum = 0;
+    let vBitrateSum = 0;
+    let aBitrateSum = 0;
+    let count = 0;
+ 
+    let rttTotal = 0;
```
[the remark this answers is not in the document]
Comment (bentcooke) [reply examples/viewer.js]: It is not...  and I have removed it. I had gone back and forth as to which to use.. RTTcurrent vs RTTtotal and ended up using RTTcurrent

File: examples/viewer.js [hunk 7]
```diff
@@ -264,3 +365,252 @@ function sendViewerMessage(message) {
         return false;
     }
 }
+
+// Functions to support DQP test and metrics
+function calcDiffTimestamp2Sec(large, small) {
```
[the remark this answers is not in the document]
Comment (bentcooke) [reply examples/viewer.js:380]: added comment for clarification

File: examples/viewer.js [hunk 8]
```diff
@@ -264,3 +373,233 @@ function sendViewerMessage(message) {
...
+            previousTime = currentTime;
+
+        } else {
+            html_str = "<table><tr><strong><FONT COLOR=RED>WAITING FOR STREAM STATS...</FONT></strong></tr></table>"
+            //Update the page
+            $('#dqp-test')[0].innerHTML =html_str; 
+            console.log("[DQP TEST] Waiting for stream stats...");
+
+            previousTime = currentTime;
+            return;
```
Comment (sirknightj) [inline examples/viewer.js:601]: minor: return is unnecessary as we are already at the end of the function

File: examples/viewer.js [hunk 9]
```diff
@@ -264,3 +373,233 @@ function sendViewerMessage(message) {
...
+        });
+
+        // Only start test and display metrics once something has been decoded
+        if ((videoDecodedFrameCount > 0) || (audioSamplesReceived > 0)){
+
+            var currentDate = new Date();
+            var currentTime = currentDate.getTime();
+
+            //timestamp start of decoded frames
+            if (statStartTime == 0) {
```
Comment (sirknightj) [inline examples/viewer.js:491]: Comparison statStartTime == 0 may cause unexpected type coercion, can we use === instead? (same in a few other places where == is used e.g. line 469)

File: examples/viewer.js (hunk 8 above)
Comment (bentcooke) [reply examples/viewer.js]: removed!

File: examples/viewer.js (hunk 9 above)
Comment (bentcooke) [reply examples/viewer.js]: fixed

Comment (sirknightj) [reply examples/viewer.js]: Few other spots: 409, 445, and 467

Comment (bentcooke) [reply examples/viewer.js]: fixed the rest with the last push

**Review by sirknightj — APPROVED** [review]

**Pull request merged**